### DISCUSSION TITLE
curl doesn't update cookie file

### DISCUSSION THREAD

**Discussion by BratSinot** [curl/curl#10117] [Q&A] [answered]
Greetings!

I have some file with cookie and I want curl to use that file and update it with new cookies from response. I try to use this:
```
curl \
  -b cookie.sub \
  -c cookie.sub \
  -d "$(jq -rc --arg ID "$ID" '.[0].id = $ID' json/handshake.json)" \
  -H 'Content-Type: application/json' \
  '[::1]:8017/notifications' |
  jq -r '.[0].clientId'
```
but curl ignores `set-cookies` from response (response have `set-cookies`, I check it through Wireshark). How can I overcome this behaviour?

cookie.sub:
```
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

#HttpOnly_::1	FALSE	/notifications/	FALSE	0	<manual set cookie name>	<cookie value>
```

Curl version:
```
✗ curl --version
curl 7.85.0 (x86_64-apple-darwin22.0) libcurl/7.85.0 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.47.0
Release-Date: 2022-08-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
```

**Comment by bagder** [comment]
As a *Q&A* this has a curious lack of question.

Your command also has a curious lack of URL so we can't reproduce.

And we don't know which curl version you use, so we cannot tell if this is an old issue you face.

There is not a lot for us to do with this.

**Reply by BratSinot** [reply]
> And we don't know which curl version you use, so we cannot tell if this is an old issue you face.
> There is not a lot for us to do with this.

Ow, right, sorry, I update the question.

**Comment by bagder** [comment]
> curl ignores set-cookies from response

This is ambiguous. You say it ignores them but there is still a cookie in the saved cookie jar?

**Reply by BratSinot** [reply]
> > curl ignores set-cookies from response
> 
> This is ambiguous. You say it ignores them but there is still a cookie in the saved cookie jar?

It is my own setted cookie. I need that:
1) request with manual set cookie;
2) response from server with `set-cookie` header;
3) new request with manual cookie and cookie from response 2;

**Reply by bagder** [reply]
Can you show us a full response that fails? I mean response headers.

**Reply by BratSinot** [reply]
> Can you show us a full response that fails? I mean response headers.

Request:
![image](https://user-images.githubusercontent.com/3810015/208479913-1611dfb8-f8e4-44f8-b714-bb710300c9e2.png)

Response:
![image](https://user-images.githubusercontent.com/3810015/208479930-68170610-896e-405c-a0a4-db600014b1b6.png)

**Comment by bagder** [comment]
Does your manually created `cookie.sub` file have a newline on the line of the only cookie?

**Comment by bagder** [comment]
I cannot reproduce.

# original cookies,txt
~~~
$ cat cookies.txt
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

#HttpOnly_::1   FALSE   /tmp/   FALSE   0       name    weird
~~~

# Get a page

use the original cookiejar, get remote cookies store in new file
~~~
$ curl '[::1]/tmp/cookie.cgi' -v -c boo.txt -b cookies.txt
*   Trying [::]:80...
* Connected to :: (::1) port 80 (#0)
> GET /tmp/cookie.cgi HTTP/1.1
> Host: [::]
> User-Agent: curl/7.86.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Mon, 19 Dec 2022 23:33:17 GMT
< Server: Apache/2.4.54 (Debian)
* Added cookie test-10117="yes" for domain ::, path /tmp/, expire 0
< Set-Cookie: test-10117=yes; HttpOnly; SameSite=None
< Upgrade: h2c
< Connection: Upgrade
< Content-Length: 4
< Content-Type: text/plain
< 
hej
* Connection #0 to host :: left intact
~~~

# check results
~~~
$ cat boo.txt 
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

#HttpOnly_::1   FALSE   /tmp/   FALSE   0       test-10117      yes
#HttpOnly_::1   FALSE   /tmp/   FALSE   0       name    weird
~~~

# curl version
~~~
$ curl -V
curl 7.86.0 (x86_64-pc-linux-gnu) libcurl/7.86.0 OpenSSL/3.0.7 zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.3 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.10.0 nghttp2/1.51.0 librtmp/2.3
Release-Date: 2022-10-26
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

**Reply by BratSinot** [reply]
Is there any debug mode in curl to check what is going on?

**Reply by BratSinot** [reply]
[wireshark.pcapng.txt](https://github.com/curl/curl/files/10266774/wireshark.pcapng.txt)
Ok here is reproduceble example:

http.go:
```go
package main

import (
	"fmt"
	"net/http"
)

func helloHandler(w http.ResponseWriter, req *http.Request) {
	// set cookie for storing token
	cookie := &http.Cookie {
	    Name: "name1",
	    Value: "Value1",
	    Secure: true,
	    HttpOnly: true,
	}
	http.SetCookie(w, cookie)

    fmt.Printf("Got cookies: `%v`\n", req.Cookies())

	fmt.Fprintf(w, "Hello World!\n")
}

func main() {
    fmt.Println(":1025")

	http.HandleFunc("/notifications", helloHandler)
	http.ListenAndServe(":1025", nil)
}
```

cookie.txt:
```
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

#HttpOnly_::1	FALSE	/notifications	FALSE	0	name0	value0
```

curl:
```
➜  curl_cookie curl -vvv -b cookie.txt -c cookie.txt '[::1]:1025/notifications'
*   Trying [::1]:1025...
* Connected to ::1 (::1) port 1025 (#0)
> GET /notifications HTTP/1.1
> Host: [::1]:1025
> User-Agent: curl/7.86.0
> Accept: */*
> Cookie: name0=value0
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Set-Cookie: name1=Value1; Path=/notifications; HttpOnly; Secure
< Date: Tue, 20 Dec 2022 08:37:36 GMT
< Content-Length: 13
< Content-Type: text/plain; charset=utf-8
<
Hello World!
* Connection #0 to host ::1 left intac
```

http.go:
```
➜  curl_cookie go run ./http.go
:1025
Got cookies: `[name0=value0]`
```

And cookie.txt doesn't change.
I also upgrade curl:
```
➜  curl_cookie curl --version
curl 7.86.0 (x86_64-apple-darwin22.1.0) libcurl/7.86.0 (SecureTransport) OpenSSL/1.1.1s zlib/1.2.11 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.4 libssh2/1.10.0 nghttp2/1.51.0 librtmp/2.3
Release-Date: 2022-10-26
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```

I also check it in `golang:1.19.4-bullseye` docker container (to get rid of possibility what it MacOS doing something), still no cookies.

**Comment by dfandrich** [comment]
The cookie being sent has the "Secure" flag set, but the server is sending it
via an insecure protocol (http). The cookie is saved fine if the server is
changed to Secure:false.

**Reply by BratSinot** [reply]
Thanks!

And can curl ignore that flag? Because if I manually add that cookie from server to file, everything work fine.

**Reply by bagder** [reply]
Since a while back, curl considers localhost, 127.0.0.1 and ::1 to be "secure". This should not be the problem, and as you can see in my example above that works, it also uses HTTP and secure.

**Reply by bagder** [reply]
Eh no my example doesn't. My mistake.

**Reply by BratSinot** [reply]
> Since a while back, curl considers localhost, 127.0.0.1 and ::1 to be "secure". This should not be the problem, and as you can see in my example above that works, it also uses HTTP and secure.

But my example in Go doesn't work.

**Comment by bagder** [comment, marked as answer]
This is issue #10120

**Reply by BratSinot** [reply]
Great, thanks!

Replaced `[::1]` by `localhost` and everything start working.
